### PR TITLE
fix(inductor): use stick-count basis in align_tensors work-division gcd

### LINKED ISSUE #3339: Stick-dim work division collapses to one core when the dim size isn't a multiple of the division
### What happened?

 pointwise op whose stick (last) dimension spans multiple sticks silently miscompiles when work division divides that stick dimension across multiple cores. Every element past the first stick is corrupted.

`align_tensors` (`torch_spyre/_inductor/views.py`) re-intersects work division's committed per-dimension core split against a variable's range via `gcd`. For a stick dimension, work division commits its core count against the **stick count** (`adjust_it_space_for_sticks`), but `align_tensors` used the **element count** as the `gcd` basis. When the element count is coprime with the committed split — e.g. 67 elements over a 2-stick / 2-core split, `gcd(2, 67) == 1` — the split collapses to a single core, so a single core now owns a fold that spans both sticks.

That collapsed layout only produces wrong output when both operands are **LX-resident** (`LX_PLANNING` on, which is the default). The LX branch of `_start_addr_data` gives every core the same buffer base address, so the single core's second-stick loads land in the wrong region. With `LX_PLANNING=0` the operands live in HBM, whose per-core addressing resolves the second stick correctly even though the split still collapses — so LX planning is a necessary condition for the miscompile, not the cause. A common pointwise producer (`neg`, `mul`, any op eligible for LX reuse) puts its result in LX under the default config, so the default path hits it readily.

Fires for any stick-dim extent `N > 64` that is coprime with the committed core split. Multiples of `64 x split` (e.g. 128, 192, 256) are clean. In multi-dimensional shapes the trigger requires the outer dims to be small enough that work division divides the stick dim rather than the outer dims.

Confirmed 1-D patterns (last dim = stick dim), where `sticks = ceil(N / 64)` and `split` is the core count work division commits to the stick dim:

| N   | sticks | split | gcd(split, N) | result       |
| --- | ------ | ----- | ------------- | ------------ |
| 64  | 1      | 1     | 1             | clean        |
| 65  | 2      | 2     | 1             | **miscompile** |
| 66  | 2      | 2     | 2             | clean        |
| 67  | 2      | 2     | 1             | **miscompile** |
| 95  | 2      | 2     | 1             | **miscompile** |
| 97  | 2      | 2     | 1             | **miscompile** |
| 127 | 2      | 2     | 1             | **miscompile** |
| 128 | 2      | 2     | 2             | clean        |
| 130 | 3      | 3     | 1             | **miscompile** |
| 160 | 3      | 3     | 1             | **miscompile** |
| 191 | 3      | 3     | 1             | **miscompile** |
| 192 | 3      | 3     | 3             | clean        |
| 193 | 4      | 4     | 1             | **miscompile** |
| 255 | 4      | 4     | 1             | **miscompile** |
| 256 | 4      | 4     | 4             | clean        |
| 257 | 5      | 5     | 1             | **miscompile** |

The miscompile fires exactly when `gcd(split, N) < split` (i.e. `N` coprime with `split`), which collapses the committed multi-core split to `gcd(split, N)` cores.

Confirmed multi-dimensional patterns (last dim = stick dim):

| shape           | stick dim divided across cores? | result       |
| --------------- | ------------------------------- | ------------ |
| `(3, 67)`       | yes                             | **miscompile** |
| `(4, 193)`      | yes                             | **miscompile** |
| `(2, 3, 67)`    | yes                             | **miscompile** |
| `(3, 5, 127)`   | yes                             | **miscompile** |
| `(2, 3, 2, 127)`| yes                             | **miscompile** |
| `(2, 129)`      | no (outer dim absorbs split)    | clean        |
| `(2, 2, 129)`   | no (outer dim absorbs split)    | clean        |
| `(2, 3, 4, 67)` | no (outer dim absorbs split)    | clean        |

### What did you expect to happen?

The compiled result to match CPU. The split that work division already certified against the stick count (2 cores, one stick each) should survive iteration-space realignment, so each stick lands on its own core.

### How can we reproduce it?

Default configuration (multi-core, `sencores=32`, `LX_PLANNING` on). The repro below uses `torch.neg` on both operands so their results are LX-resident; a plain `x + y` of two graph inputs keeps the operands in HBM and does not fire.

Two settings independently mask it:

- `SENCORES=1` — single core, so there is no split to collapse (removes the collapse).
- `LX_PLANNING=0` — operands live in HBM, whose per-core addressing resolves the second stick correctly (removes the LX single-base addressing). The split still collapses under this setting.

```python
import torch

def fn(x, y):
    return torch.neg(x) + torch.neg(y)

# 67 elements = 2 sticks (64 + 3), coprime with the 2-core split.
# Integer values <= 1024 are exact in device fp16, so this is a real mismatch.
n = 67
x = (torch.arange(n) % 251).to(torch.float16)
y = ((torch.arange(n) + 100) % 251).to(torch.float16)

got = torch.compile(fn, dynamic=False)(x.to("spyre"), y.to("spyre")).cpu()
ref = fn(x, y)
torch.testing.assert_close(got, ref)   # fails: elements 64..66 corrupted
```

### Any environmental details we need to know?

- torch-spyre: `origin/main` at `0883cb6`
- PyTorch: `2.11.0`

### Anything else we need to know?

Root cause is a pure basis mismatch: `work_division.py` already uses the stick count as the split basis for stick vars (`enumerate_work_division_candidates`), then `align_tensors` re-intersects the same split against the element count — a basis its own upstream never uses for a stick var. Fix: feed the stick-count basis (`ceil(elems / elems_per_stick)`) into the `gcd` for the stick var's segment; non-stick vars keep the element basis.

### PR DESCRIPTION
**Pull request by yoheiueda**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:
A pointwise op whose stick (last) dimension spans multiple sticks silently miscompiles when work division divides that stick dimension across multiple cores — every element past the first stick is corrupted.

`align_tensors` (`torch_spyre/_inductor/views.py`) re-intersects work division's committed per-dimension core split against a variable's range via `gcd`. For a stick dimension, work division commits its core count against the **stick count** (`adjust_it_space_for_sticks`), but `align_tensors` used the **element count** as the `gcd` basis. When the element count is coprime with the committed split — e.g. 67 elements over a 2-stick / 2-core split, `gcd(2, 67) == 1` — the split collapses to a single core, so one core now owns a fold spanning both sticks.

The fix feeds the stick-count basis (`ceil(elems / elems_per_stick)`) into the `gcd` for the stick var's reused first segment; synthetic sub-segments and all non-stick vars keep the element basis, so genuine multi-segment non-stick splits are unchanged.

Fixing that `gcd` collapse exposed a second latent defect. The DXP backend compiler cannot apply coordinate masking to a dimension split across cores; such masking is applied to a dim that is padded, reduced, and the stick dim. `enumerate_work_division_candidates` already rejected splits dividing such a dim, but the greedy default path (`_default_split` → `multi_dim_iteration_space_split`) did not — it could commit a split entirely on a padded reduction stick dim, aborting the backend compile. Previously the element-basis `gcd` accidentally collapsed that split to one core (e.g. `gcd(8, 511) == 1`), masking the gap; once the stick-count fix keeps the split alive, the abort surfaces. This PR extracts the shared predicate `coordinate_mask_blocked_vars()` and has both paths consult it, so the greedy path drops such dims from its reduction candidates. This fixes the backend `SIGABRT` on `test_t_2d_127x131_lx_planning_reduction` and `test_copy_roundtrip_1d_padded_lx_planning_reduction`.

Adds a regression test (`test_binary_op_stick_crossing_last_dim`) exercising a pointwise binary op whose stick dim spans multiple sticks with a coprime extent, across 1D–4D shapes and 2/3/4-stick spans. Registered in both the base (`test_inductor_ops_pointwise_config.yaml`) and LX-planning (`test_inductor_ops_lx_pointwise_config.yaml`) CI configs.

#### Which issue(s) this PR is related to:

Fixes #3339

#### Special notes for your reviewer:

The collapsed layout only produces wrong output when both operands are **LX-resident** (`LX_PLANNING` on, the default). With `LX_PLANNING=0` the operands live in HBM and the same collapsed split compiles correctly — so LX planning is a *necessary condition* for the miscompile, not the cause. The repro uses `torch.neg` on both operands so their results are LX-resident; a plain `x + y` of two graph inputs keeps them in HBM and does not fire. `SENCORES=1` also masks it (no split to collapse).

**Why LX is affected and HBM is not.** The difference is entirely in how `_start_addr_data` (`torch_spyre/_inductor/codegen/compute_ops.py`) computes each core's start address. For an HBM tensor it adds a per-core work-slice offset, so every core reads from the region it actually owns:

```python
# HBM branch: per-core base = start_address + this core's work-slice offset
return {
    f"[{c}, 0, 0]": str(
        tensor.start_address
        + core_idx_to_slice_offset(
            tensor, core_id_to_wk_slice[str(c)], sdsc_spec.work_slices
        )
        * num_bytes(tensor.data_format)
    )
    for c in range(sdsc_spec.num_cores)
}
```

For an LX tensor the `core_idx_to_slice_offset` term is dropped entirely — every core is handed the *same* base address:

```python
# LX branch: every core gets the identical base, no per-core slice offset
if "lx" in tensor.allocation:
    return {
        f"[{c}, 0, 0]": str(tensor.start_address)
        for c in range(sdsc_spec.num_cores)
    }
```

Normally that is fine, because work division puts each stick on its own core and each core's kernel only touches its own stick relative to that shared base. But once the split collapses to a single core, that one core's fold spans *both* sticks off the single LX base `{0, 128}` — and the second stick's loads land in the wrong region. The HBM path is immune to the same collapse because its per-core offset is recomputed per read edge; even a single-core fold resolves the second stick against a base that already folds in the work-slice offset.

This PR targets the collapse rather than the LX addressing: keeping the split at two cores lands each stick on its own core, so the single-core cross-stick fold never forms and the LX single-base addressing is never exercised. The underlying LX limitation (one per-core base address, no per-access offset) is a separate backend-side capability gap.

#### Does this PR introduce a user-facing change?

Fixes a silent miscompile for pointwise ops whose stick (last) dimension has an extent coprime with the multi-core work-division split.

#### Additional note:

The parameterized binary-op shapes never covered this case: they carry a stick-aligned last dim and place odd counts on outer dims, so a coprime extent never landed on the stick dim.

### CODE REVIEW

**Comment by yoheiueda** [comment]
@cyang49 @tardieu @dgrove-oss 

Could you please take a look? I discovered this issue while working on #3110.

**Review by dgrove-oss — COMMENTED** [review]
Summary

Fixes a silent miscompile: align_tensors re-derived per-dim core splits using the raw element count as the gcd basis for a stick (last) dimension, but work division commits splits against the stick count. When the element count is coprime with the committed split (e.g. 67 elements over a 2-stick/2-core split), the split silently collapsed to one core, corrupting every element past the first stick for LX-resident operands. Fixing that collapse exposed a second latent bug — the greedy _default_split path could commit a core split on a stick dim that also carries padding + reduction, which the backend can't coordinate-mask, causing a SIGABRT. The PR extracts the existing enumerate_work_division_candidates predicate into a shared coordinate_mask_blocked_vars() helper and has the greedy path consult it too.

Review

I traced the eps/basis arithmetic in views.py:631-635 against adjust_it_space_for_sticks in work_division.py:330-336 — both use the identical (elements + elems_per_stick - 1) // elems_per_stick ceiling-division to convert an element count to a stick count, confirming the fix's basis is correct and consistent with the rest of the codebase's stick-count convention. The coordinate_mask_blocked_vars() extraction is a byte-for-byte lift of the previous inline predicate in enumerate_work_division_candidates (no behavior change there), and the three call sites (enumerate_work_division_candidates, work_distribution_pass, _cost_model_divide_op) all compute coord_vars/reduction_vars/blocked identically before calling _default_split. The new blocked filter in _default_split only prunes candidates the greedy distributor may pick — it can't cause wrong output, only reduced parallelism in the worst case. Well-targeted regression test covering 1D–4D shapes and 2/3/4-stick spans, wired into both base and LX-planning CI configs; CI is fully green.

BLOCKERs

None.

SUGGESTIONs

- The coord_vars = {...}; reduction_vars = [...]; blocked = coordinate_mask_blocked_vars(...) triplet is now duplicated verbatim across three call sites (work_division.py in enumerate_work_division_candidates, work_distribution_pass, and _cost_model_divide_op). Consider factoring it into a small helper (e.g. _reduction_vars_and_blocked(output_td, it_space_adjusted, stick_vars, it_space)) to avoid the three copies drifting apart later.

Verdict

APPROVE

**Comment by dgrove-oss** [comment]
merging blocked on review from @tardieu

**Review by tardieu — APPROVED** [review]
lgtm. Thanks for the detailed analysis of the multiple issues involved.
